### DISCUSSION TITLE
About os.dupterm readinto() method

### DISCUSSION THREAD

**Discussion by jd3096-mpy** [micropython/micropython#11600] [Using MicroPython] [answered]
https://github.com/boochow/FBConsole

Like this repo,we can redirect repl by using os.dupterm,by the method write(),we can get the repl print and redirect to screen.
But how to change input source like martix keyboard?

In docs,it says:
Duplicate or switch the MicroPython terminal (the REPL) on the given [stream](https://docs.micropython.org/en/latest/reference/glossary.html#term-stream)-like object. The stream_object argument must be a native stream object, or derive from io.IOBase and implement the readinto() and write() methods. The stream should be in non-blocking mode and readinto() should return None if there is no data available for reading.

So,how to use readinto() the change repl input source from UART0 RX to other source?

**Comment by jimmo** [comment]
You need to write a stream object that returns input from your keyboard device. It can just ignore anything that's written to it.

The idea is that the REPL is duplicated to multiple streams (e.g. UART, keyboard, framebuffer, BLE, etc). Each one just implements what makes sense -- UART & BLE do both input and output, framebuffer does just output, keyboard does just input.

**Reply by jd3096-mpy** [reply]
I got what you mean.But I don't know how to write the readinto() method.
Can you give a simple example code like push a button to input letter 'a' to the repl?

https://github.com/boochow/FBConsole/issues/1

**Reply by jimmo** [reply, marked as answer]
Here is an example (based on https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_repl.py) that runs on a Pyboard and generates an "A" each time the pyboard's button is pressed.

Your code will need to do something different to get the key events, but ultimately implement read/readinto to return characters in a similar way.

```py
import io
import pyb
import os

_MP_STREAM_POLL = const(3)
_MP_STREAM_POLL_RD = const(0x0001)

class ButtonStream(io.IOBase):
    def __init__(self, c):
        pyb.Switch().callback(self._press)
        self._n = 0
        self._c = c

    def _press(self):
        self._n += 1
        if hasattr(os, "dupterm_notify"):
            os.dupterm_notify(None)

    def read(self, sz=None):
        buf = bytearray(self._n if sz is None else sz)
        self.readinto(buf)
        return buf

    def readinto(self, buf):
        n = min(len(buf), self._n)
        for i in range(n):
            buf[i:i+1] = self._c
        self._n -= n
        return None if n == 0 else n

    def ioctl(self, op, arg):
        if op == _MP_STREAM_POLL and self._buf:
            return _MP_STREAM_POLL_RD
        return 0

    def write(self, buf):
        pass

os.dupterm(ButtonStream(b'A'))
```

**Reply by jd3096-mpy** [reply]
Thanks a lot, that's very helpful.

**Reply by rkompass** [reply]
@jimmo I'm just wondering if it is correct that `ButtonStream.read()` does not return anything?
Or is there a `return buf` missing?

**Reply by jimmo** [reply]
@rkompass yes you're right, thanks! I've updated the snippet above.

FWIW, `read` isn't actually necessary here, only `readinto` is used. You could leave it off and dupterm would still work.

dupterm uses the internal stream protocol interface, which is implemented by IOBase to use the user-defined readinto/write/ioctl methods only.
